distogram/__init__.py

```diff
@@ -1,240 +1,203 @@
+    ]
+    h.min_diff = min(diffs)
 
     return diffs
 
 
-def _linear_index(bins, value):
-    ratio = float((value - bins[0][0]) / (bins[-1][0] - bins[0][0]))
-    index = round(ratio * len(bins))
-    if index > 0 and index < (len(bins) - 1):
-        if value > bins[index][0] and value < bins[index+1][0]:
-            return index+1
-        if value > bins[index-1][0] and value < bins[index][0]:
-            return index
-
-    return 0
-
-
-def _search_in_place_index(h, new_value, index):
-    bins = h.bins
-
+def _search_in_place_index(h: Distogram, new_value: float, index: int) -> int:
     if h.diffs is None:
         h.diffs = _compute_diffs(h)
 
-    min_diff = h.min_diff
-    i_bin = None
-
     if index > 0:
-        diff1 = _weighted_diff(h, (new_value, 1), bins[index-1])
-        diff2 = _weighted_diff(h, bins[index], (new_value, 1))
-        if diff1 < diff2:
-            diff = diff1
-            i_bin = index-1
-        else:
-            diff = diff2
-            i_bin = index
+        diff1 = _weighted_diff(h, (new_value, 1), h.bins[index - 1])
+        diff2 = _weighted_diff(h, h.bins[index], (new_value, 1))
+
+        i_bin, diff = (index - 1, diff1) if diff1 < diff2 else (index, diff2)
+
+        return i_bin if diff < h.min_diff else -1
 
-        if diff < min_diff:
-            return i_bin
-        else:
-            return -1
     return -1
 
 
-def update(h, value, count=1):
-    '''Adds a new element to the distribution.
+def update(h: Distogram, value: float, c: int = 1) -> Distogram:
+    """ Adds a new element to the distribution.
 
     Args:
         h: A Distogram object.
         value: The value to add on the histogram.
-        count: [Optional] The number of times that value must be added.
+        c: [Optional] The number of times that value must be added.
 
     Returns:
         A Distogram object where value as been processed.
-    '''
-    bins = h.bins
+    """
     index = 0
-    if len(bins) > 0:
-        if value <= bins[0][0]:
+    if len(h.bins) > 0:
+        v0, _ = h.bins[0]
+        vl, _ = h.bins[-1]
+
+        if value <= v0:
             index = 0
-        elif value >= bins[-1][0]:
+        elif value >= vl:
             index = -1
         else:
-            if len(bins) >= h.bin_count:
-                index = _linear_index(bins, value)
-            if index == 0:
-                index = bisect_left(bins, (value, 1))
+            index = bisect_left(h.bins, (value, 1))
 
-    if index > 0 and len(bins) >= h.bin_count:
-        in_place_index = _search_in_place_index(h, value, index)
-        if in_place_index > 0:
-            h = _trim_in_place(h, value, count, in_place_index)
+        vi, fi = h.bins[index]
+        if vi == value:
+            h.bins[index] = (vi, fi + c)
             return h
 
-    if len(bins) > 0:
-        bin = bins[index]
-        if bin[0] == value:
-            bin = (bin[0], bin[1]+count)
-            h.bins[index] = bin
```

Comment:
`bin` is a built-in function to convert something to a binary representation. This runs the risk of renaming the `bin` function for a user.